File: cpp/tests/internal/rotation.cpp
// File: rotation.cpp
// Project: internal
// Created Date: 26/11/2023
// Author: Shun Suzuki
// -----
// Last Modified: 02/12/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/internal/controller.hpp>
#include <autd3/internal/geometry/rotation.hpp>
#include <autd3/link/audit.hpp>
#include <numbers>

#include "utils.hpp"

TEST(Internal, Angle) {
  ASSERT_NEAR((90.0 * autd3::internal::geometry::deg).to_radian(), std::numbers::pi / 2, 1e-6);
  ASSERT_NEAR((std::numbers::pi / 2 * autd3::internal::geometry::rad).to_radian(), std::numbers::pi / 2, 1e-6);
}

static inline autd3::internal::Controller<autd3::link::Audit> open_with_rotation(const autd3::internal::Quaternion& q) {
  return autd3::internal::ControllerBuilder()
      .add_device(autd3::internal::geometry::AUTD3(autd3::internal::Vector3::Zero()).with_rotation(q))
      .open_with_async(autd3::link::Audit::builder())
      .get();
}

TEST(Internal, WithRotation) {
  using autd3::internal::Vector3;
  using autd3::internal::geometry::deg;
  using autd3::internal::geometry::EulerAngles;

  {
    const auto autd = open_with_rotation(EulerAngles::from_zyz(90.0 * deg, 0.0 * deg, 0.0 * deg));
    ASSERT_NEAR_VECTOR3(Vector3::UnitY(), autd.geometry()[0][0].x_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(-Vector3::UnitX(), autd.geometry()[0][0].y_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitZ(), autd.geometry()[0][0].z_direction(), 1e-6);
  }
  {
    const auto autd = open_with_rotation(EulerAngles::from_zyz(0.0 * deg, 90.0 * deg, 0.0 * deg));
    ASSERT_NEAR_VECTOR3(-Vector3::UnitZ(), autd.geometry()[0][0].x_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitY(), autd.geometry()[0][0].y_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitX(), autd.geometry()[0][0].z_direction(), 1e-6);
  }
  {
    const auto autd = open_with_rotation(EulerAngles::from_zyz(0.0 * deg, 0.0 * deg, 90.0 * deg));
    ASSERT_NEAR_VECTOR3(Vector3::UnitY(), autd.geometry()[0][0].x_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(-Vector3::UnitX(), autd.geometry()[0][0].y_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitZ(), autd.geometry()[0][0].z_direction(), 1e-6);
  }
  {
    const auto autd = open_with_rotation(EulerAngles::from_zyz(0.0 * deg, 90.0 * deg, 90.0 * deg));
    ASSERT_NEAR_VECTOR3(Vector3::UnitY(), autd.geometry()[0][0].x_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitZ(), autd.geometry()[0][0].y_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitX(), autd.geometry()[0][0].z_direction(), 1e-6);
  }
  {
    const auto autd = open_with_rotation(EulerAngles::from_zyz(90.0 * deg, 90.0 * deg, 0.0 * deg));
    ASSERT_NEAR_VECTOR3(-Vector3::UnitZ(), autd.geometry()[0][0].x_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(-Vector3::UnitX(), autd.geometry()[0][0].y_direction(), 1e-6);
    ASSERT_NEAR_VECTOR3(Vector3::UnitY(), autd.geometry()[0][0].z_direction(), 1e-6);
  }
}

File: dotnet/cs/src/Rotation.cs
/*
 * File: Rotation.cs
 * Project: src
 * Created Date: 26/11/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

using AUTD3Sharp.NativeMethods;

#if UNITY_2018_3_OR_NEWER
using UnityEngine;
using Quaternion = UnityEngine.Quaternion;
#else
using Quaternion = AUTD3Sharp.Utils.Quaterniond;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp
{
    public struct Angle
    {
        public static Angle FromDegree(float_t degree) => new Angle(degree / 180 * AUTD3.Pi);
        public static Angle FromRadian(float_t radian) => new Angle(radian);

        public float_t Radian { get; }

        public class UnitRadian
        {
            internal UnitRadian() { }
            public static Angle operator *(float_t a, UnitRadian b) => FromRadian(a);
        }
        public class UnitDegree
        {
            internal UnitDegree() { }
            public static Angle operator *(float_t a, UnitDegree b) => FromDegree(a);
        }

        public static class Units
        {
            public static UnitRadian Rad { get; } = new UnitRadian();
            public static UnitDegree Deg { get; } = new UnitDegree();
        }

        private Angle(float_t value)
        {
            Radian = value;
        }
    }

    public static class EulerAngles
    {
        public static Quaternion FromZYZ(Angle z1, Angle y, Angle z2)
        {
            unsafe
            {
                float_t* rot = stackalloc float_t[4];
                NativeMethodsBase.AUTDRotationFromEulerZYZ(z1.Radian, y.Radian, z2.Radian, rot);
                return new Quaternion(rot[1], rot[2], rot[3], rot[0]);
            }
        }
    }
}